### ISSUE TITLE
Release build not working

### ISSUE THREAD

**Issue by FurryFur** [wjakob/nanogui#258]
Nanogui crashes when I try to build anything with it in release mode. I get an access violation error. This seems to be happening on Screen construction. Specifically, it seems like it is crashing when it tries to copy construct the mCaption member variable. It only crashes in release mode, everything works fine in debug. This is really problematic since I need to be able to build a release mode executable for my project. This is the line of code that seems to be crashing. 
```c
screen = new Screen(Vector2i(500, 700), "NanoGUI test");
```
This seems to be the same or related issue to #257, however, I have opened a new one since the title of that issue doesn't accurately describe the problem or its severity.

**Comment by svenevs** [comment]
I feel like these both may be Eigen related.  Can you build using `d8903a47fd873ebe279c58f6f9bdb4ac15eaa38c`?

```
$ cd /path/to/fresh/nanogui/clone
$ git checkout d8903a47fd873ebe279c58f6f9bdb4ac15eaa38c
```

?

**Comment by FurryFur** [comment]
I can't get that version to build, it complains about a bunch of failed template specializations for Eigen.

**Comment by FurryFur** [comment]
Not sure how you guessed it had to do with Eigen, but it seems like you were right. I've tracked the issue down to Eigen's cwiseMax function. It crashes in the window.cpp file in the preferredSize function when it tries to call cwiseMax in the return statement with an Access Violation exception. 

It seems to be the same issue described in this [StackOverflow thread](https://stackoverflow.com/questions/42954418/eigen-cwisemin-cwisemax-access-violation-in-visual-studio-2017-release-build/42962860#42962860). Although I am using Visual Studio 2015 x64 and not 2017 x86, the issue described seems to be almost identical, including the fact that if I replace the vector components with constants it resolves the issue. But as soon as you put variables in the vectors to be used in the cwiseMax function, it crashes.

**Comment by FurryFur** [comment]
Tried updating Eigen to the latest version but it didn't make any difference.

**Comment by FurryFur** [comment]
I was able to build libraries using Visual Studio 2017 which I can use in my 2015 projects. So seems like this might be a compiler issue.

**Comment by svenevs** [comment]
> Not sure how you guessed it had to do with Eigen, but it seems like you were right.

Eigen was recently updated in NanoGUI, and the other recent changes wouldn't have caused something like this :wink:

> I was able to build libraries using Visual Studio 2017 which I can use in my 2015 projects. So seems like this might be a compiler issue.

So to be clear, the latest version of `nanogui` on `master` works in VS 2017 without any changes on your part?  There is a slightly hidden problem here.  Even if NanoGUI were to remove all calls to `cwiseMax`, the user has to commit to one version of Eigen for both NanoGUI and their own code.  `cwiseMax` has potential to get called implicitly, so it basically means that the user will likely encounter a bug in their own code (if using Eigen elsewhere).

If VS2017 works with latest `master`, I think the best option would be to update the docs and state clearly this is a bug.

If you're feeling motivated, you could try using the latest NanoGUI with an **older** version of Eigen.  NanoGUI used to work fine with VS2015, the version of Eigen being used before the update was [here](https://github.com/libigl/eigen/tree/c34a9130bc585b288703bd9716d7efae194974e2).

You could just clone that locally.  Directions for custom Eigen haven't been added to the official docs yet, I explained how [over here](https://github.com/wjakob/nanogui/issues/226#issuecomment-307696049).  Unfortunately there are still reasons to have to stick with VS2015, depending on the project's other dependencies (besides NanoGUI).

**Comment by FurryFur** [comment]
Yes it works without modification with latest master. The only thing I did differently was to have cmake output VS 2017 project files. I think adding it to the docs would be an acceptable solution for now. I'm not sure what else you could do, other than report the bug to Eigen and/or Microsoft.

**Comment by svenevs** [comment]
Righto, thanks for investigating all of this!  I'm actually reworking the docs currently.  I decided to go-for-gold and redo `entypo.h` which got a little complicated, but I'll make sure the compilation docs get updated for this as well as how to incorporate an alternative Eigen to use with NanoGUI.

**Comment by fbergama** [comment]
I confirm that in Release mode it crashes with VS2015. Any clue yet? I cannot upgrade to VS2017 anytime soon :(

**Comment by hanke100321** [comment]
i find a fork of this repository [https://github.com/libigl/nanogui](https://github.com/libigl/nanogui) which can be work in vs2015

**Comment by FurryFur** [comment]
I've just reported this as a bug to the Eigen developers via email. Unfortunately for some reason, you can't submit bug reports to their bug tracker without an account and account registration is locked... So I had to send it via email.

**Comment by wjakob** [comment]
I think that VS2015 **x64** *should* work. The issue might be a missing ``EIGEN_MAKE_ALIGNED_OPERATOR_NEW`` statement in some class that is misaligned as a result. Was anybody able to extract some kind of a usable backtrace?

**Comment by chhtz** [comment]
@FurryFur I did create an Eigen-bugzilla account for you, but you need to report the bug yourself (even though it looks very much like a MSVC issue, there may be a workaround we can do -- but I'm really not an MSVC expert).

@wjakob At least the minimal example he sent us does not involve anything which could be fixed by `EIGEN_MAKE_ALIGNED_OPERATOR_NEW` (and it did not involve alignment at all)

**Comment by FurryFur** [comment]
I just submitted it now

**Comment by FurryFur** [comment]
For the record, even this small piece of code causes a crash:
```c
#include <iostream>
#include <time.h>

#include <Eigen\Core>

int main() {
	int i = rand();
	Eigen::Vector3i a(i, i, i);
	Eigen::Vector3i b(10, 10, 10);
	Eigen::Vector3i min = a.cwiseMin(b);

	std::cout << min << std::endl;

	return 0;
}
```
So I don't think it has anything to do with NanoGUI.

**Comment by svenevs** [comment]
Ok windows users, #260 should finally be ready.  It adds docs on how to manually specify Eigen.  For right now, [the fake docs are here](http://fake-nanogui.readthedocs.io/en/latest/compilation.html#nanogui-and-eigen) (don't bookmark that though, that's just a testing site).

Please try using an **older** version of Eigen with VS2015.  If you're just using NanoGUI without a specific parent project, [NanoGUI used to use this commit on the Eigen Repo](https://github.com/libigl/eigen/tree/c34a9130bc585b288703bd9716d7efae194974e2).

**Comment by ggael** [comment]
I don't have MSVC to try to understand this issue, so could someone publish a complete backtrace to see where it fails? I've read in some  SO comments that it's failing in windows.h, so it could be that MSVC is messing with its min/max macros. To check that, could someone try to replace in Eigen/src/Core/MathFunctions.h:

    template<typename T>
    EIGEN_DEVICE_FUNC
    EIGEN_ALWAYS_INLINE T mini(const T& x, const T& y)
    {
      EIGEN_USING_STD_MATH(min);
      return min EIGEN_NOT_A_MACRO (x,y);
    }

by 

    template<typename T>
    EIGEN_DEVICE_FUNC
    EIGEN_ALWAYS_INLINE T mini(const T& x, const T& y)
    {
      return x<y ? x : y;
    }

Same for maxi just below.

If that does not do the trick, then I'm clueless.

**Comment by twMr7** [comment]
Just found out that using a dynamic-size vector in PR #275 could just be a workaround, there are plenty of fixed-sized cwiseMax() and cwiseMin() in codes. For those who want to get back to work quickly on Visual Studio 2105, the trick to use a dynamic-size vector shall do.
